src/app/obfboard.ts

```diff
@@ -1,25 +1,56 @@
+import {
+  IsString,
+  IsNotEmpty,
+  IsOptional,
+  IsDefined,
+  IsInt,
+  IsUrl,
+  ValidateNested,
+  validateSync,
+  ValidationError,
+} from 'class-validator';
 import { ImageResolver } from './image-resolver';
 import { SoundResolver } from './sound-resolver';
+import { FatalOpenVoiceFactoryError, ErrorCodes } from './errors';
+import { Check2DArray, OneOf } from './custom-validation';
 
 export class Grid {
+
+  @IsDefined()
+  @IsInt()
   rows: number;
+
+  @IsDefined()
+  @IsInt()
   columns: number;
+
+  @Check2DArray('columns', 'rows')
   order: string[][];
 
   deserialize(input: any): Grid {
     this.rows = input.rows;
     this.columns = input.columns;
     this.order = input.order.map(row => row.map(cell => cell || cell === 0 ? String(cell) : undefined));
-
     return this;
   }
 }
 
 export class LoadBoardAction {
+
+  @IsNotEmpty({ message: 'Load board action id must be specified'})
+  @IsString({ message: 'Load board action id must be a string'})
```

Comment:
According to the OBF spec (and examples, and our code usage) load board action id is not necessary, or relevant, or used, and is possibly misleading

Reply:
However, load board must specify one of url, dataUrl or path, so we should probably use the OneOf validator on it